fw/http.c

```diff
@@ -2983,6 +2983,52 @@ tfw_h1_set_loc_hdrs(TfwHttpMsg *hm, bool is_resp, bool from_cache)
 	return 0;
 }
 
+/*
+ * Rewrite HTTP/1 "PURGE" method to "GET" directly inside a request SKB.
+ */
+static int
+tfw_h1_rewrite_purge_to_get(struct sk_buff *head)
+{
+	const char *q = "GET";
+	char *p;
+	struct skb_shared_info *si;
+	struct sk_buff *skb = head;
+	unsigned int f, z;
+	int ret;
+
+	/* Chop two bytes from the beginning of SKB data. */
+	ret = ss_skb_chop_head_tail(head, head, 2, 0);
+	if (ret)
+		return ret;
+
+	/* Possible if somehow we already sent a response  */
+	BUG_ON(!head);
+
+	do {
+		p = skb->data;
+		z = skb_headlen(skb);
+		while (z--) {
+			*p++ = *q++;
+			if (!*q)
+				return 0;
+		}
+		si = skb_shinfo(skb);
+		for (f = 0; f < si->nr_frags; ++f) {
+			p = skb_frag_address(&si->frags[f]);
+			z = skb_frag_size(&si->frags[f]);
+			while (z--) {
+				*p++ = *q++;
+				if (!*q)
+					return 0;
+			}
+		}
+		skb = skb->next;
+	} while(skb != head);
```

Comment:
I think we should use the standard `skb_store_bits()` here

Reply:
I'm sorry, how is this supposed to work? `skb_store_bits` tries to copy into the linear area, failing that - into the fragments, failing that - into the fragment list which *I think* we unroll earlier in a socket callback. Otherwise it returns EFAULT. `skb_store_bits` isn't going to walk over our skb list.

Reply:
Also, I thought you didn't want us to use `memcpy` just to set three bytes?

Reply:
> skb_store_bits isn't going to walk over our skb list.

Yeah, good point.

> Also, I thought you didn't want us to use memcpy just to set three bytes?

No, `memcpy()` is good since compiler can inline it, but our own `memcpy_fast()` is bad for small copies since it requires a far call. But the current implementation is good as well.